src/ripple/conditions/prefix_sha256.h

```diff
@@ -0,0 +1,204 @@
+            prefix_.begin(),
+            prefix_.end(),
+            b.begin());
+
+        std::copy (
+            subpayload.begin(),
+            subpayload.end(),
+            out);
+
+        return b;
+    }
+
+    condition_t
+    condition() const override
+    {
+        auto const& sc = subcondition();
+        auto const blob = to_blob (sc);
+
+        auto const payloadSize = oer::predict_octetstring_size (
+            sc.fulfillment_length);
+
+        // FIXME: we can reduce copying here by incrementally
+        // hashing:
+        Buffer b (blob.size() +
+            oer::predict_octetstring_size (prefix_.size()));
+
+        auto out = oer::encode_octetstring (
+            prefix_.size(),
+            prefix_.begin(),
+            prefix_.end(),
+            b.begin());
+
+        std::copy (
+            blob.begin(),
+            blob.end(),
+            out);
+
+        sha256_hasher h;
+        h (b.data(), b.size());
+
+        condition_t cc;
+        cc.type = type();
+        cc.requires = requires();
+        cc.fulfillment_length = prefix_.size() + payloadSize;
+        cc.fingerprint = static_cast<sha256_hasher::result_type>(h);
+
+        return cc;
+    }
+
+    std::uint16_t
+    type () const override
+    {
+        return condition_prefix_sha256;
+    }
+
+    std::uint32_t
+    requires () const override
+    {
+        return
+            feature_sha256 |
+            feature_prefix |
+            subfulfillment_->requires();
+    }
+
+    bool
+    validate (Slice const& m) const override
+    {
+        if (!subfulfillment_)
+        {
+            std::cerr << "Failed because of no subf!\n";
+            return false;
+        }
+
+        // Prepend the prefix to the message:
+        Buffer b (prefix_.size() + m.size());
+
+        if (prefix_.size())
+            std::memcpy (b.data(), prefix_.data(), prefix_.size());
+
+        if (m.size())
+            std::memcpy (b.data(), m.data(), m.size());
+
+        return subfulfillment_->validate (b);
+    }
+
+    fulfillment_t const&
+    subfulfillment () const
+    {
+        return *subfulfillment_;
+    }
+
+    condition_t
+    subcondition () const
+    {
+        return subfulfillment_->condition();
+    }
+
+    bool
+    parse_payload (Slice s) override
+    {
+        // The payload consists of the prefix, followed by
+        // a subfulfillment:
+
+        auto iter = s.begin ();
+
+        std::size_t len;
+
+        std::tie (iter, len) = oer::decode_length (
+            iter, s.end());
+
+        if (len != 0)
+        {
+            std::memcpy (prefix_.alloc (len), iter, len);
+            std::advance (iter, len);
+        }
+
+        s += std::distance (s.begin(), iter);
+
+        // The remaining bytes in the slice are a fulfillment
+        // so we parse it as such. If we can, then we've
+        // succeeded.
+        subfulfillment_ = load_fulfillment (s);
+
+        return static_cast<bool>(subfulfillment_);
```

Comment:
If loading the subfulfillment fails we should "clean up our mess" by clearing `prefix_`. We didn't document the object invariants, but a empty prefix/empty subfullments seems valid (empty state), but valid prefix empty subfulfillment seems invalid.